### PR TITLE
Fix resuming the tqdm progress bar 

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

The changes in #12889 have caused #13124. 

This PR reverts #12889 (except tests) and adds the fix to call `tqdm.initial = 0`, which prevents from tqdm incorrectly computing the estimates for ETA and iterations/s. All tests added in #12889 remain as they are not affected by reverting this. 

Since this is a visual bug, and the tqdm produces these estimates at runtime, I don't really know how to test this. Suggestions by reviewers are welcome!

<details>
  <summary>Repro example used to debug:</summary>
  

```py

import os
import shutil
from time import sleep

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        sleep(.1)
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        print()
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("val_loss", loss)
        print()

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 32), batch_size=2)

    if os.path.exists("lightning_logs"):
        shutil.rmtree("lightning_logs")

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        enable_model_summary=False,
        enable_progress_bar=True,
        callbacks=ModelCheckpoint(monitor="train_loss", save_top_k=-1, every_n_train_steps=1),
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=train_data)

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=3,
        enable_model_summary=False,
        enable_progress_bar=True,
        callbacks=ModelCheckpoint(monitor="train_loss", save_top_k=-1, every_n_train_steps=1),
    )
    trainer.fit(
        model, train_dataloaders=train_data, ckpt_path="lightning_logs/version_0/checkpoints/epoch=0-step=3.ckpt"
    )


if __name__ == "__main__":
    run()

```

</details>

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @tchaton @rohitgr7 @awaelchli @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 1]
```diff
@@ -375,9 +379,7 @@ def convert_inf(x: Optional[Union[int, float]]) -> Optional[Union[int, float]]:
...
 
 
-def _update_n(bar: _tqdm, current: int, refresh_rate: int) -> None:
+def _update_n(bar: _tqdm, value: int) -> None:
     if not bar.disable:
-        total = bar.total
-        leftover = current % refresh_rate
-        advance = leftover if (current == total and leftover != 0) else refresh_rate
-        bar.update(advance)
+        bar.n = value
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/progress/tqdm_progress.py:384]: do you think this is enough to make sure the time estimates are correctly updated?

last time I checked `update` does more than just setting the progress bar counter.
https://github.com/tqdm/tqdm/blob/4f208e72552c4d916aa4fe6a955349ee8b2ed353/tqdm/std.py#L1212-L1277

Comment (awaelchli) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:384]: My understanding is that update() was never the problem, even in https://github.com/Lightning-AI/lightning/pull/12889. The problem is in resetting when a new epoch begins. The running estimate that the bar has needs to be reset, otherwise it becomes negative and it would take a few iterations for it to converge to the new value. The only way to achieve this I found is by setting initial = 0. 

[Here](https://github.com/tqdm/tqdm/blob/4f208e72552c4d916aa4fe6a955349ee8b2ed353/tqdm/std.py#L441-L445) is where the computation for the rate happens.
Right below that, the remaining time gets formatted based on the values computed in the rate, hence this is directly affected. 

Note that in your PR in #12889 the value for n is not correct after resuming. The progress bar starts at 0 even though it should start at the step where the checkpoint was saved. This is now fixed by reverting here.

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 1 above)
Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:384]: I see. At that time, the negative estimates started coming after long training, but that fix introduced this issue where training is resumed.

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 2]
```diff
@@ -254,12 +254,13 @@ def on_train_start(self, *_: Any) -> None:
     def on_train_epoch_start(self, trainer: "pl.Trainer", *_: Any) -> None:
         total_batches = self.total_batches_current_epoch
         self.main_progress_bar.reset(convert_inf(total_batches))
+        self.main_progress_bar.initial = 0
```
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/progress/tqdm_progress.py:257]: maybe we should reinitialize it to`initial=reloaded value` which is `self.train_batch_idx + self._val_processed` for main progress bar.

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 1 above)
Comment (rohitgr7) [inline src/pytorch_lightning/callbacks/progress/tqdm_progress.py:384]: and keep this as it is? to make sure `miniters` is considered if set by user using tqdm options.
I don't know a lot about tqdm internals but manually setting just the value of `n` skips the update for `last_print_n` and `last_print_t`.

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:257]: Setting it to that leads to negative values in the progress bar. 
After one epoch, these counters are outdated.

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:384]: The previous way with update() would increase the bar by increments, but this is not correct when resuming with our counters. We can look for a better solution but this is the only one I found that worked.

Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:384]: just want to make sure the old reported issues don't come up again since there's no way to test this in the CI to avoid regression. Will explore more on this.. let's merge this then for now.

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:257]: I don't think so they will be outdated since they are reset.

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:257]: Print the values and see for yourself. Doesn't get reset. Hence I can't use it as "initial value". 

image

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:384]: Yes I saw the issues that the previous PR fixed. I checked [this](https://github.com/Lightning-AI/lightning/issues/12875) yesterday and it was also caused by not resetting tqdm.initial to 0.

File: src/pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/progress/tqdm_progress.py:257]: ok.. I thought its batch_start hook since it's the case with others.. and there it's 0 after each epoch.

**Review by kaushikb11 — APPROVED** [review]
